KAFKA-7824: Require member.id for initial join group request

abbccdda opened the pull request:
Hey there,

this is my draft implementation of [KIP-394](https://cwiki.apache.org/confluence/display/KAFKA/KIP-394%3A+Require+member.id+for+initial+join+group+request), requiring member id when the new join group request was received. The current blocker was on the concurrent unit tests, and I figure it would be good to post the implementation at the same time to have some parallel review.

Basically the major refactor is that we extract all the unknown member handling logic as a separate function called `doUnknownJoinGroup` , and define empty MemberMetadata struct called `unknownMemberMetadata ` which will be inserted into the members list when we receive a valid unknown member join group request. This mock object will be evicted after the rebalance timeout.

Let me know if the high level implementation makes sense here when you got time @hachikuji @guozhangwang @MayureshGharat @mjsax @stanislavkozlovski 

Also we created a separate JIRA: https://issues.apache.org/jira/browse/KAFKA-7824 for this PR, since 7610 is closed with another PR.

Thanks a lot!

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
add unknown member metadata

Commit:
fix unit test

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/JoinGroupRequest.java:
```diff
@@ -76,9 +76,17 @@
+    /**
+     * The version number is bumped to indicate that client needs to issue a second join group request under first try
+     * with UNKNOWN_MEMBER_ID.
+     * TODO: Define v4 schema to include `group.instance.id` for KIP-345.
```
nit: we can leave this out. we can just bump the version to v5 for KIP-345. Some people run from trunk.

abbccdda replied:
done

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -548,6 +548,13 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
                 future.raise(error);
             } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
                 future.raise(new GroupAuthorizationException(groupId));
+            } else if (error == Errors.MEMBER_ID_REQUIRED) {
```
Can we add a unit test to verify this path?

abbccdda replied:
The changes I made to the #joinGroup function in GroupCoordinatorTest is actually serving as the unit test for this logic.

hachikuji replied:
Yeah, I was referring to the testing of the client-side logic. It should be possible to modify one of the existing tests in either `AbstractCoordinatorTest` or `ConsumerCoordinatorTest`.

abbccdda replied:
Sound good！

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
+    val firstMemberWithEmptyProtocolOrProtocolType = group.is(Empty) && (protocols.isEmpty || protocolType.isEmpty)
+    val notMatch = groupProtocolNotSupportedByNewMember || firstMemberWithEmptyProtocolOrProtocolType
+    if (notMatch) {
+      responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
```
nit: it is a little surprising to have this method respond to the callback. I think it would be clearer to leave this as a simple check and let the caller send the response.

abbccdda replied:
done

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -112,6 +112,11 @@ private object GroupMetadata {
+  val unknownMemberMetadata = new MemberMetadata("", "", "", "", 0, 0, "", List[(String, Array[Byte])]())
+  def isUnknownMember(member: MemberMetadata): Boolean = member == unknownMemberMetadata
+
+  // TODO: decide if ids should be predictable or random
```
nit: since we're touching this, how about we get rid of this TODO?

abbccdda replied:
I guess you mean we should just remove it?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -207,22 +212,42 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+                             protocolType: String,
+                             protocols: Set[String]): Unit = {
+    if (members.isEmpty)
+      this.protocolType = Some(protocolType)
```
From the original code, but we should probably save any mutation until we have validated the rest of the metadata.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -112,6 +112,11 @@ private object GroupMetadata {
       Stable -> Set(CompletingRebalance),
       PreparingRebalance -> Set(Stable, CompletingRebalance, Empty),
       Empty -> Set(PreparingRebalance))
+  val unknownMemberMetadata = new MemberMetadata("", "", "", "", 0, 0, "", List[(String, Array[Byte])]())
```
Hmm.. I think it would be clearer to track the new members in a separate set (say `pendingMembers`). It's a bit dangerous to mix int the pending members with sentinel metadata in the main collection. Consider, for example, the check below `hasAllMembersJoined` which compares the number of members (which may include some unknown) with the number that are awaiting join. That check no longer seems to work.

abbccdda replied:
Yep make sense, I'm on it.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
+        // for new members. If the new member is still there, we expect it to retry.
+        completeAndScheduleNextExpiration(group, member, NewMemberJoinTimeoutMs)
       }
+      maybePrepareRebalance(group, s"Adding new member $newMemberId")
```
In the case that `requireKnownMemberId` is true, I wonder if we should hold off transitioning to `PreparingRebalance` until the member has joined with new memberId?

stanislavkozlovski replied:
That sounds like the way to go, doesn't it? We can't be sure that a rebalance will happen until we receive the second request with the memberId in that case

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
+      // If member id required, register the member with unknown metadata,
+      // and send back a response to call for another join group request with allocated member id.
+      if (requireKnownMemberId) {
+        group.addUnknownMember(newMemberId, group.groupId, protocolType, protocols)
```
I am wondering if we should call `completeAndScheduleNextHeartbeatExpiration`. This would ensure that the pending member will eventually get timed out if the consumer never rejoins.

stanislavkozlovski replied:
I also think we might want to call it - that call should satisfy the KIP's statement:

>  We also handle the registered member id eviction through session timeout so that the pre-allocation map will not grow indefinitely, although the map size should be trivial since we only store a random generated id.

right?

abbccdda replied:
When I start the implementation, I actually realize that it's probably simpler just to keep the `pendingMembers` map as a simple check without enforcing purgatory. The reason is that the size of the map is trivial and will be evicted after rebalance timeout, which I feel could be more clean and efficient (saving purgatory space in the meantime). What do you think?

stanislavkozlovski replied:
The rebalance timeout is client-controlled, right? It's in the JoinGroup request and it comes from `max.poll.interval.ms` as far as I know. Users can set that to Integer.MAX_INT (as they have with that original JIRA that spawned the discussions and KIP-389) and fill this up to infinity (at least it's just a string and not MemberMetadata)

In my understanding, that's what we're partly fixing with this PR.

abbccdda replied:
Yep, +1. It's necessary to handle the rebalance timeout case here, thanks for pointing out!

hachikuji reviewed (commented):
@abbccdda Thanks for the patch! I left a few initial comments. Note, by the way, that you need to rebase and there are probably checkstyle issues.

abbccdda commented:
@hachikuji Great thanks for the comments Jason！

Commit:
move unknown metadata out of members

Commit:
debug

stanislavkozlovski commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,22 +117,98 @@ class GroupCoordinator(val brokerId: Int,
+    val newMemberId = clientId + "-" + GroupMetadata.generateMemberIdSuffix
+    var group = groupManager.getGroup(groupId).orNull
+    if (group == null) {
       // only try to create the group if the group is not unknown AND
```
here we're creating a group if the group is unknown and the member id is unknown, which goes opposite to the comment (saying that the group **should not** be unknown)

Is this a stale comment or am I not interpreting "unknown group" correctly?

abbccdda replied:
I think this is stale comment. Let me rephrase it to be correct. You could double check the trunk to see that it actually doesn't make sense (correct way: group is unknown && member id is unknown)

stanislavkozlovski commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -335,15 +341,36 @@ class GroupCoordinatorTest extends JUnitSuite {
 
   @Test
-  def testHeartbeatWrongCoordinator() {
+  def testJoinGroupUnknownConsumerDeaGroup() {
```
nit: typo - "Dead"

stanislavkozlovski commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -335,15 +341,36 @@ class GroupCoordinatorTest extends JUnitSuite {
+    EasyMock.reset(replicaManager)
+    responseFuture = sendJoinGroup(groupId, joinGroupResult.memberId, "invalidProtocolType", protocols, requireKnownMemberId = true)
+    joinGroupResult = Await.result(responseFuture, Duration(DefaultRebalanceTimeout + 1, TimeUnit.MILLISECONDS))
+    assertEquals(Errors.INCONSISTENT_GROUP_PROTOCOL, joinGroupResult.error)
```
In this case, would a third request with the correct protocol finish the join? Maybe it wouldn't hurt to test that as well

abbccdda replied:
Sounds good!

stanislavkozlovski commented:
Thanks @abbccdda, this is good work

abbccdda commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -1682,8 +1710,20 @@ class GroupCoordinatorTest extends JUnitSuite {
-    val responseFuture = sendJoinGroup(groupId, memberId, protocolType, protocols, rebalanceTimeout, sessionTimeout)
+                        rebalanceTimeout: Int = DefaultRebalanceTimeout,
+                        requireKnownMemberId: Boolean = DefaultRequireKnownMemberIdInJoinGroup): JoinGroupResult = {
+    var responseFuture = sendJoinGroup(groupId, memberId, protocolType, protocols, sessionTimeout, rebalanceTimeout, requireKnownMemberId)
```
@hachikuji This is the part that I tested the new logic.

stanislavkozlovski commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -241,9 +250,10 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
 
   def currentState = state
 
+  // TODO: deprecate the kick out logic for normal member's rejoin failure within rebalance timeout.
```
In my opinion, it's better to just create a JIRA rather than a TODO comment, since those generally don't get addressed (e.g the one you removed with this PR) and don't contain as much context

abbccdda replied:
I plan to address this in the next PR (KIP-345). In case I'm not available for future work, others could easily pick up when seeing the comments :)

hachikuji replied:
We usually prefer JIRAs over comments. It is better for discoverability and prevents cluttering up the code.

abbccdda replied:
Thanks for the info. Let me remove this comment.

Commit:
add pending member to pugatory

Commit:
remove purgatory key change

Commit:
refactor addMemberAndRebalance to take in the member directly

abbccdda commented:
@hachikuji @stanislavkozlovski I have handled the pending member removal within heartbeat purgatory, fixed unit tests and added some corresponding tests on the client side. Feel free to take another look!

Commit:
add getGeneration for unit test

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -729,6 +738,14 @@ protected synchronized Generation generation() {
+     * Return generation info regardless of the group state.
+     * @return the current generation info
+     */
+    protected synchronized Generation getGeneration() {
```
nit: we could probably make this use default access. Also, we usually drop the `get` from getters. So this could just be `generation()`.

abbccdda replied:
There is already a `generation()` function within the context. I could try to rename the existing one with `generationOnGroupStable()`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -294,7 +295,9 @@
             "election so the offsets cannot be guaranteed to be monotonically increasing",
-            OffsetNotAvailableException::new);
+            OffsetNotAvailableException::new),
+    MEMBER_ID_REQUIRED(79, "Consumer needs to have a valid member id before actually entering group",
```
Can we remove the mention of the consumer specifically? This protocol is also used in connect. Maybe we can just say "The group member needs to have a..."?

abbccdda replied:
Sounds good!

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:
```diff
@@ -176,6 +176,24 @@ public void testJoinGroupRequestMaxTimeout() {
+        String memberId = "memberId";
+        int generation = 10;
+
+        mockClient.prepareResponse(joinGroupFollowerResponse(generation, memberId, "leaderId", Errors.MEMBER_ID_REQUIRED));
```
I think the simple way to do this is to prepare two responses. For the second one, we can provide a `RequestMatcher` which validates that the correct `memberId` is provided in the request. Then we can just use `ensureActiveGroup`.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -239,11 +240,18 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }
   }
 
+  def isPendingMember(memberId: String): Boolean = pendingMembers.contains(memberId) && !has(memberId)
```
Is it ever possible that a memberId is in both collections?

abbccdda replied:
In the current logic, no, adding the check here just to ensure no one accidentally break the handling logic.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,25 +117,68 @@ class GroupCoordinator(val brokerId: Int,
+        responseCallback(joinError(JoinGroupRequest.UNKNOWN_MEMBER_ID, Errors.INCONSISTENT_GROUP_PROTOCOL))
+      } else {
+        val newMemberId = clientId + "-" + group.generateMemberIdSuffix
+        val newMember = new MemberMetadata(newMemberId, group.groupId, clientId, clientHost, rebalanceTimeoutMs,
```
Hmm.. I don't think we want to hold a reference to the consumer's state until they rejoin. For the purpose of expiration, we should just need the memberId. Maybe we could replace the `MemberMetadata` field inside `DelayedHeartbeat` with the memberId? Then we could get rid of the `pending` flag inside `MemberMetadata` and just rely on the `pendingMembers` collection.

abbccdda replied:
We do need to have a full MemberMetadata, check out `shouldKeepAlive` function.

I guess we could add a new field called pendingMemberId to the DelayedHeartbeat struct and set MemberMetadata to null. How does that sound?

hachikuji replied:
Yes, my suggestion above was to drop `MemberMetadata` from `DelayedHeartbeat` and use `memberId` instead

abbccdda replied:
If we drop `MemberMetadata`, how do we know if we need to keep the member alive?

abbccdda replied:
I think I figure out a way. Let me update the diff.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -145,83 +189,70 @@ class GroupCoordinator(val brokerId: Int,
+          sessionTimeoutMs, protocolType, protocols)
+
+        addMemberAndRebalance(rejoiningMember, group, responseCallback)
+        group.removePendingMember(memberId)
```
I wonder if we could move this `GroupMetadata.add(member)`. That would us to encapsulate the pending state a little better.

abbccdda replied:
I think we could move `group.removePendingMember(memberId)` into the `addMemberAndRebalance`, because it serves no harm to the existing join logic. We are simply trying to remove a non-exist key.

stanislavkozlovski commented:
LGTM! Failing test seems unrelated

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -145,83 +189,70 @@ class GroupCoordinator(val brokerId: Int,
         responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
       } else {
         group.currentState match {
-          case Dead =>
```
It's a little annoying that we had to make this match non-exhaustive. This will probably generate a warning. Maybe we can add default unexpected case. For example:

```
case unexpected =>
  warn("Attempt to rebalance existing member of group $groupId in unexpected state $unexpected")
  responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
```
I think this could also be used for the `Empty` case (see my comment below).

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -145,83 +189,70 @@ class GroupCoordinator(val brokerId: Int,
-        responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
-      } else if (group.is(Empty) && (protocols.isEmpty || protocolType.isEmpty)) {
-        //reject if first member with empty group protocol or protocolType is empty
+      if (group.currentState == Dead) {
```
nit: `group.is(Dead)`

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -145,83 +189,70 @@ class GroupCoordinator(val brokerId: Int,
-              // if the member id is unknown, register the member to the group
-              addMemberAndRebalance(rebalanceTimeoutMs, sessionTimeoutMs, clientId, clientHost, protocolType,
-                protocols, group, responseCallback)
+            val member = group.get(memberId)
```
With the changes here, I think the `Empty` case shouldn't be possible. We know we have one member at least.

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

Commit:
address Jason's comment

Commit:
refactor heartbeat purgatory to include pending member id

Commit:
augment coordinator test

abbccdda commented:
@hachikuji @stanislavkozlovski could I get another review for this?

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -723,12 +732,20 @@ protected synchronized void markCoordinatorUnknown(boolean isDisconnected) {
+     * Return generation info regardless of the group state.
+     * @return the current generation info
+     */
+    protected synchronized Generation generation() {
```
This seems to be used only in tests, maybe it is worth clarifying in the javadoc?

abbccdda replied:
Sounds good!

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -294,7 +295,9 @@
             "election so the offsets cannot be guaranteed to be monotonically increasing",
-            OffsetNotAvailableException::new);
+            OffsetNotAvailableException::new),
+    MEMBER_ID_REQUIRED(79, "The group member needs to have a valid member id before actually entering group",
```
nit `entering group` -> `entering the group`/`entering a consumer group`

stanislavkozlovski commented on core/src/main/scala/kafka/coordinator/group/DelayedHeartbeat.scala:
```diff
@@ -25,12 +25,13 @@ import kafka.server.DelayedOperation
 private[group] class DelayedHeartbeat(coordinator: GroupCoordinator,
                                       group: GroupMetadata,
+                                      pendingMemberId: String,
                                       member: MemberMetadata,
```
Since this can be null, would it be better if we used an `Option<MemberMetadata`>?

abbccdda replied:
I will be removing the member from heartbeat purgatory.

stanislavkozlovski commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -844,17 +886,21 @@ class GroupCoordinator(val brokerId: Int,
-      if (!member.shouldKeepAlive(heartbeatDeadline)) {
+      if (pendingMemberId != JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+        // Clean out pending member from purgatory through the member given session timeout.
+        group.removePendingMember(pendingMemberId)
```
Maybe a `DEBUG` level log would be good here

Commit:
addr comm

Commit:
fix test

abbccdda commented:
@hachikuji FYI, I removed the member metadata from DelayedHeartbeat. Member status could be retrieved through `group.get(memberId)`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -548,6 +550,13 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
+            } else if (error == Errors.MEMBER_ID_REQUIRED) {
+                // Broker requires a concrete member id to be allowed to join the group. Update generation and member id
+                // and send another join group request in next cycle.
+                AbstractCoordinator.this.generation = new Generation(joinResponse.generationId(),
```
I'm slightly uncomfortable with the additional complexity that comes from having an incomplete `Generation` object. At a minimum, we probably need to change the checks which rely on `Generation.NO_GENERATION`. For example, see `maybeLeaveGroup`. Perhaps we can add an `isValid` method to `Generation` which only returns true if the generationId is valid?

abbccdda replied:
IIUC, you are suggesting that we should narrow down the check to only looking at whether given generation object has a valid `generation number`. I add `isValid()` here only to check if this is true by comparing with invalid generation id -1. Let me know if this is what we should do.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -723,12 +732,20 @@ protected synchronized void markCoordinatorUnknown(boolean isDisconnected) {
+     * Return generation info regardless of the group state. Only using in unit tests.
+     * @return the current generation info
+     */
+    protected synchronized Generation generation() {
```
Can we drop to default access? I don't think sub-classes will want to touch this.

abbccdda replied:
Changing keyword to `final`. Is this what you are suggesting?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -723,12 +732,20 @@ protected synchronized void markCoordinatorUnknown(boolean isDisconnected) {
      * @return the current generation or null if the group is unjoined/rebalancing
      */
-    protected synchronized Generation generation() {
+    protected synchronized Generation generationOnGroupStable() {
```
`AbstractCoordinator` is an unofficially exposed API. We generally don't mind breaking it, but we try to avoid it if we can. I wonder if we could leave `generation()` as it was and introduce alternative getters with a different name? Maybe if we just had a `memberId()` and `hasValidGenerationId()`, that would be good enough for testing?

abbccdda replied:
That makes sense. I reverted the refactor.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:
```diff
@@ -176,6 +176,41 @@ public void testJoinGroupRequestMaxTimeout() {
+        final String memberId = "memberId";
+        final int generation = 10;
+
+        mockClient.prepareResponse(joinGroupFollowerResponse(generation, memberId, "leaderId", Errors.MEMBER_ID_REQUIRED));
```
I'm not sure it makes sense to use a valid generation and leaderId here. The coordinator actually returns generationId=0 and leaderId="" at the moment for the MEMBER_ID_REQUIRED case.

As an aside, we should probably fix `GroupCoordinator.joinError()`. I can't think of a good reason why we should return 0 instead of `GroupCoordinator.NoGeneration`. Can we do that in this patch?

abbccdda replied:
Yep, fixing it.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,25 +117,65 @@ class GroupCoordinator(val brokerId: Int,
-            val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
-            doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-          }
+      return
```
nit: explicit returns are frowned upon in Scala. Was there a reason not to keep the `else` branch?

abbccdda replied:
Not really, it's just a coding style that I'm not aware :) Thanks for the info!

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -116,25 +117,65 @@ class GroupCoordinator(val brokerId: Int,
+    group.inLock {
+      if (group.is(Dead)) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
```
nit: `this is likely` -> `it is likely`?

abbccdda replied:
fixed all the comment occurrence.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -844,16 +886,31 @@ class GroupCoordinator(val brokerId: Int,
-      if (member.shouldKeepAlive(heartbeatDeadline) || member.isLeaving)
-        forceComplete()
-      else false
+      if (isPending || !group.has(memberId))
```
Could we use `isPendingMember` and get rid of the flag?

abbccdda replied:
It's a bit hard. The flag I setup here is to let Heartbeat purgatory be aware whether this is a delayed operation for pending member or for active member. The `pendingMembers` struct could be changed before the delayed operation timeouts, if that happens, `isPendingMember` couldn't know whether this *was* a pending member or not, thus assuming active members timing out and remove member in accident.

hachikuji replied:
Should we have logic to complete the DelayedOperation when the pending member joins?

hachikuji replied:
I think I'd suggest we change this logic to the following:

```scala
if (isPending)
  group.has(memberId)
else {
...
```
In other words, for a pending member, the heartbeat is considered "complete" when the member rejoins. Without this logic, we always have to await the pending member heartbeat expiration even if the member has joined. This ensures that we cleanup these delayed operations as soon as the member joins.

hachikuji reviewed (commented):
@abbccdda Thanks for the updates. I think we're close. I left a few more comments.

Commit:
fix comments and refactor return

Commit:
fix no generation comparison

abbccdda commented:
I double checked that test failings are not related. It is not reproduced locally.

abbccdda commented:
@hachikuji Do you want to take another look? Thanks!

hachikuji commented:
@abbccdda Yes, I will try to look this afternoon. In the meantime, can you double-check the test failures? Typically this is symptomatic of a test case which is leaking a client. I'd make sure each new test case is closing all resources.

abbccdda commented:
@hachikuji cool, sounds good!

abbccdda commented:
rebased to the latest

Commit:
Fix concurrent access in abstract coordinator

abbccdda commented:
@hachikuji Thanks for the suggestion! I double checked the test and realized there was a concurrent modification error in the KafkaConsumer test since I forgot to put the generation modification within a synchronized block. Let's see if this run works!

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:
```diff
@@ -176,6 +176,41 @@ public void testJoinGroupRequestMaxTimeout() {
+        assertTrue(consumerClient.poll(future, mockTime.timer(REQUEST_TIMEOUT_MS)));
+        assertEquals(Errors.MEMBER_ID_REQUIRED.message(), future.exception().getMessage());
+        assertTrue(coordinator.rejoinNeededOrPending());
+        assertEquals(memberId, coordinator.memberId());
```
I think this check is redundant given the validation in the request matcher above. Perhaps we can remove this assertion and the `memberId()` method.

abbccdda replied:
Sounds good!

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:
```diff
@@ -176,6 +176,41 @@ public void testJoinGroupRequestMaxTimeout() {
+        assertEquals(Errors.MEMBER_ID_REQUIRED.message(), future.exception().getMessage());
+        assertTrue(coordinator.rejoinNeededOrPending());
+        assertEquals(memberId, coordinator.memberId());
+        assertTrue(coordinator.hasValidGenerationId(generation));
```
Maybe `hasMatchingGenerationId` would be a better name since the generation is not valid? Or perhaps we can remove the argument and return true iff the generationId is not -1?

abbccdda replied:
Sure I will rename the function. I do feel a comparison shall be more handy.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -729,6 +740,27 @@ protected synchronized Generation generation() {
+     * @return true if the two ids are matching.
+     */
+    final synchronized boolean hasValidGenerationId(int generationId) {
+        return generation.generationId == generationId;
```
Missing null check?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -548,6 +550,15 @@ public void handle(JoinGroupResponse joinResponse, RequestFuture<ByteBuffer> fut
+                // Broker requires a concrete member id to be allowed to join the group. Update generation and member id
+                // and send another join group request in next cycle.
+                synchronized (AbstractCoordinator.this) {
+                    AbstractCoordinator.this.generation = new Generation(joinResponse.generationId(),
```
Rather than relying on the returned generation, perhaps we can use `OffsetCommitRequest.DEFAULT_GENERATION_ID` explicitly?

abbccdda replied:
Sounds good！

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -844,19 +885,32 @@ class GroupCoordinator(val brokerId: Int,
-        info(s"Member ${member.memberId} in group ${group.groupId} has failed, removing it from the group")
-        removeMemberAndUpdateGroup(group, member, s"removing member ${member.memberId} on heartbeat expiration")
+      if (isPending) {
+        debug(s"Pending member $memberId gets cleaned out from purgatory after the member given session timeout.")
```
Maybe more concise?

> Pending member $memberId has been removed after session timeout expiration

hachikuji reviewed (commented):
Thanks for the updates, just a few more comments.

Commit:
remove member id helper

abbccdda commented:
@hachikuji So far I don't see the necessity of completing the DelayedOperation for pending members because they only take up very small amount of memory. It should be more efficient to let it naturally kick out IMO.

I will keep checking why the unit tests are failing.

abbccdda commented:
@hachikuji currently the constant failing test is `testMultiConsumerRoundRobinAssignment`:
```
16:26:00     java.lang.AssertionError: Did not get valid initial assignment for partitions [topic2-3, topic1-2, topic2-0, topic2-5, topic1-4, topic2-7, topic2-1, topic2-6, topic1-0, topic1-3, topic2-4, topic1-1, topic2-2]
16:26:00         at kafka.utils.TestUtils$.fail(TestUtils.scala:356)
16:26:00         at kafka.utils.TestUtils$.waitUntilTrue(TestUtils.scala:766)
16:26:00         at kafka.api.PlaintextConsumerTest.validateGroupAssignment(PlaintextConsumerTest.scala:1772)
16:26:00         at kafka.api.PlaintextConsumerTest.testMultiConsumerRoundRobinAssignment(PlaintextConsumerTest.scala:864)
16:26:00 
```
Is there any suggestion for further debugging? I couldn't reproduce the issue on local.

Commit:
extend wait time a bit for round-trip

abbccdda commented:
@hachikuji I think it might be due to a second round trip takes longer time to react. I'm increasing the wait time to 20 seconds to see if it works.

abbccdda commented:
@hachikuji Seems that increasing time doesn't help here. Do you have any suggestion to reproduce the bug on local?

Commit:
fix state to unjoined to avoid concurrent modification

abbccdda commented:
@hachikuji  Updates: I managed to reproduce the bug on local, which is resource leak for not closing heartbeat thread. Right now the fix is to explicitly set member state to UNJOINED for AbstractorCoordinator. Although I haven't figured out which part of the code gets leaked due to member state, I will try multiple times to ensure the tests are good.

Commit:
increase group initial rebalance delay to avoid concurrent modification

Commit:
rebase to add valid versions

abbccdda commented:
@hachikuji Finally all tests are passing! Could you take another look for this patch? Sorry for taking so long, I do see the difficulty in fixing multi-threaded tests :)

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -52,7 +52,7 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
   this.serverConfig.setProperty(KafkaConfig.GroupMinSessionTimeoutMsProp, "100") // set small enough session timeout
   this.serverConfig.setProperty(KafkaConfig.GroupMaxSessionTimeoutMsProp, "30000")
-  this.serverConfig.setProperty(KafkaConfig.GroupInitialRebalanceDelayMsProp, "0")
+  this.serverConfig.setProperty(KafkaConfig.GroupInitialRebalanceDelayMsProp, "10")
```
Was this needed in the end?

abbccdda replied:
The tests `testMultiConsumerSessionTimeoutOnStopPolling` and `testMultiConsumerSessionTimeoutOnClose` are having 3 consumers join the group concurrently. With one more bounce, it is easy to see that the last 2 consumers join late which triggers second rebalance that takes longer time and test could timeout. Adding 10ms for initial.join.delay could mitigate the problem.

abbccdda commented:
@hachikuji Hey Jason, could you take another look when you got time? Hope to get this done soon :) Thanks!

hachikuji commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -60,6 +60,8 @@ class GroupCoordinatorTest extends JUnitSuite {
   val GroupMaxSessionTimeout = 10 * 60 * 1000
   val DefaultRebalanceTimeout = 500
   val DefaultSessionTimeout = 500
+  val DefaultRequireKnownMemberIdInJoinGroup = true
```
nit: since these are just used as default parameter values, we can probably remove them

abbccdda replied:
Sounds good

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -253,11 +260,17 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
 
   def canRebalance = GroupMetadata.validPreviousStates(PreparingRebalance).contains(state)
 
+  def protocolNotMatch(memberProtocolType: String,
```
Would it be possible to consolidate this logic into `supportsProtocols`? Also, I think it might be simpler to do a simple `if (is(Empty))` check and split the two cases.

abbccdda replied:
Simplified the logic by using isEmpty to check the two cases. And converged two checks.

hachikuji reviewed (commented):
Just a couple more comments.

Commit:
address jason's comments

Commit:
merge protocolnotmatch with supportsprotocol func

abbccdda commented:
I addressed the latest comments, now let's cross the finger for letting the build pass :) @hachikuji

Commit:
fix unit test

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

hachikuji commented:
@abbccdda Before I merge, can you create a separate JIRA? We ended up closing 7610 with a separate patch. Also, can you update the PR description?

abbccdda commented:
SGTM! I will get a separate JIRA then.

abbccdda commented:
@hachikuji The new JIRA ticket is created. I updated the PR title to match the JIRA, is this what you mean by updating PR description?

hachikuji commented:
@abbccdda Yep, that works. Thanks.

abbccdda commented:
@hachikuji Thanks a lot Jason!
